workflow/controller/operator.go

```diff
@@ -1906,16 +1862,79 @@ func (woc *wfOperationCtx) executeTemplate(ctx context.Context, nodeName string,
 				// unexpected behavior and is a bug.
 				panic("bug: GetLockName should not return an error after a call to TryAcquire")
 			}
+			woc.log.Infof("Not Acquire lockname: %s", lockName)
 			return woc.markNodeWaitingForLock(node.Name, lockName.EncodeName()), nil
 		} else {
 			woc.log.Infof("Node %s acquired synchronization lock", nodeName)
 			if node != nil {
 				node = woc.markNodeWaitingForLock(node.Name, "")
 			}
+			// Set this value to check that this node is using synchronization, and acquire the lock.
+			unlockedNode = true
 		}
 
 		woc.updated = woc.updated || wfUpdated
 	}
+
+	// If memoization is on, check if node output exists in cache
+	if processedTmpl.Memoize != nil {
+		// Apply memoize only when node is nil or node is using synchronization (mutex) that acquiring the lock.
+		// This additional condition will make the cache correctly works after synchronization.\
```

Comment:
I'm thinking maybe modify the comment to something like:
`// Check memoization cache if the node is about to be created, or was created in the past but is only now allowed to run due to acquiring a lock`
What do you think?